src/persist/src/indexed/mod.rs

```diff
@@ -331,51 +317,27 @@ impl<U: Buffer, L: Blob> Indexed<U, L> {
             }
         }
 
-        let entry = BufferEntry { updates };
-        let mut entry_bytes = Vec::new();
-        unsafe { abomonation::encode(&entry, &mut entry_bytes) }
-            .expect("write to Vec is infallible");
-        let seqno = self.buf.write_sync(entry_bytes)?;
-        Ok(seqno)
-    }
-
-    /// Atomically moves all writes currently in the buffer into the future.
-    fn drain_buf(&mut self) -> Result<(), Error> {
-        let mut updates_by_id: HashMap<Id, Vec<(SeqNo, (Vec<u8>, Vec<u8>), u64, isize)>> =
-            HashMap::new();
-        let desc = self.buf.snapshot(|seqno, buf| {
-            // Skip updates not in advance of the last sequence number we have observed.
-            if seqno < self.futures_seqno_upper {
-                return Ok(());
-            }
-            let mut buf = buf.to_vec();
-            let (entry, remaining) = unsafe { abomonation::decode::<BufferEntry>(&mut buf) }
-                .ok_or_else(|| Error::from(format!("invalid buffer entry")))?;
-            if !remaining.is_empty() {
-                return Err(format!("invalid buffer entry").into());
-            }
-            for (id, updates) in entry.updates.iter() {
-                // iter and cloned instead of append because I don't have a mental
-                // model of what's safe with abomonation.
-                updates_by_id.entry(*id).or_default().extend(
-                    updates.iter().map(|((key, val), ts, diff)| {
-                        (seqno, (key.clone(), val.clone()), *ts, *diff)
-                    }),
-                );
-            }
+        let prev = self.serialize_meta();
 
-            Ok(())
-        })?;
+        let mut updates_by_id: HashMap<Id, Vec<((Vec<u8>, Vec<u8>), u64, isize)>> = HashMap::new();
 
-        // If there's nothing in the buffer we can exit early because there's
-        // nothing left to do.
-        if desc.start == desc.end {
-            debug_assert!(updates_by_id.is_empty());
-            debug_assert_eq!(self.futures_seqno_upper, desc.end);
-            return Ok(());
+        for (id, updates) in updates.into_iter() {
+            updates_by_id.entry(id).or_default().extend(updates);
         }
 
-        let prev = self.serialize_meta();
+        // Give each write a unique, incrementing sequence number, and use
+        // futures_seqno_upper to track the sequence number of the next write.
+        let ret = self.futures_seqno_upper;
+        self.futures_seqno_upper = SeqNo(ret.0 + 1);
+
+        // This range represents the [lower, upper) of sequence numbers assigned
+        // to this write.
+        // TODO: do we still need sequence numbers? This will make more sense
```

Comment:
yeah, i think we'll want them to be able to data in buffer that didn't make it to trace